guia/genericos.md

```diff
@@ -1 +1,324 @@
+
+```swift
+func swapTwoInts(inout a: Int, inout _ b: Int)
+func swapTwoValues<T>(inout a: T, inout _ b: T)
+```
+
+A versão genérica da função usa um _nome de tipo reservado_ no local do  (chamado `T`, neste caso), ao invés do _verdadeiro_ nome do tipo (tal como `Int`, `String` ou `Double`). O nome de tipo reservado não diz nada sobre o que `T` deve ser, mas ele diz que ambos `a` e `b` devem ser do mesmo tipo `T`, seja lá o que `T` representa. O tipo verdadeiro a ser usado no lugar de `T` será determinado cada vez que a função `swapTwoValues(_:_:)` é chamada.
+
+A outra diferença é que o nome da função genérica (`swapTwoValues(_:_:)`) é seguida pelo nome de tipo reservado (`T`) dentro dos sinais de maior/menor (`<T>`). Isso diz ao Swift que `T` é um nome de tipo reservado dentro da definição da função `swapTwoValues(_:_:)`. Como `T` é um nome de tipo reservado, Swift não procura por um tipo real chamado `T`.
+
+A função `swapTwoValues(_:_:)` pode agora ser chamada da mesma forma que `swapToInts(_:_:)`, exceto que agora pode ser passado para ela dois valores de quaisquer tipo, contanto que ambos os valores sejam do mesmo tipo entre si. Cada vez que `swapTwoValues(_:_:)` é chamada, o tipo `T` é inferido dos tipos dos valores passados para a função.
+
+Nos dois exemplos abaixo, `T` é inferido como `Int` e `String` respectivamente:
+
+```swift
+var someInt = 3
+var anotherInt = 107
+swapTwoValues(&someInt, &anotherInt)
+// someInt vale agora 107, e anotherInt vale agora 3
+ 
+var someString = "hello"
+var anotherString = "world"
+swapTwoValues(&someString, &anotherString)
+// someString vale agora "world", e anotherString vale agora "hello”
+```
+
+> NOTA
+>
+> A função `swapTwoValues(_:_:)` definida acima é inspirada pela função genérica `swap`, que faz parte da biblioteca padrão de Swift, e está automaticamente disponível para você usar em em seus aplicativos. Se você precisa do comportamento de `swapTwoValues(_:_:)` em seu código, use a função `swap(_:_:_)` disponível na biblioteca padrão de Swift, ao invés de usar sua própria implementação.
+
+### Tipos de Parâmetros
+
+No exemplo de `swapTwoValues(_:_:)` acima, o nome de tipo reservado `T` é um exemplo de _tipo de parâmetro_. Um tipo de parâmetro especifica e nomeia um tipo reservado, e é escrito imediatamente depois do nome da função, entre um par de sinais de menor e maior. (assim como `<T>`).
+
+Uma vez que você especifique um tipo de parâmetro, , você pode usa-lo para definir os tipos dos parâmetros de uma função (tais como `a` e `b` na função `swapTwoValues(_:_:)`), ou como o valor devolvido por uma função, ou como anotação do tipo dentro do corpo de uma função. Em cada caso, o tipo de parâmetro é substituído com o tipo _verdadeiro_ sempre que a função é chamada. (no exemplo acima de `swapTwoValues(_:_:)`, `T` foi substituído com `Int` na primeira vez que a função foi chamada, e foi substituído por `String` na segunda vez que foi chamado.)
```

Comment:
E na última, o `no exemplo acima de...` acho que pode começar com maíuscula no início da sentença entre parênteses.